File: bada_ruby/lib/bada/transformer/tensor.rb
# frozen_string_literal: true

module Bada
  module Transformer
    # Minimal pure-Ruby tensor helpers (matrices as Array<Array<Float>>).
    # Everything the manifold-gauge transformer needs, no external deps.
    module Tensor
      module_function

      def zeros(rows, cols)
        Array.new(rows) { Array.new(cols, 0.0) }
      end

      # Deterministic weight matrix in [-scale, scale] from a seeded xorshift PRNG.
      def randmat(rows, cols, rng, scale: 0.5)
        Array.new(rows) { Array.new(cols) { (rng.next_float * 2.0 - 1.0) * scale } }
      end

      # A x B  (a: m x k, b: k x n) -> m x n
      def matmul(a, b)
        m = a.length
        k = b.length
        n = b[0].length
        out = Array.new(m) { Array.new(n, 0.0) }
        i = 0
        while i < m
          ai = a[i]
          oi = out[i]
          t = 0
          while t < k
            aval = ai[t]
            if aval != 0.0
              bt = b[t]
              j = 0
              while j < n
                oi[j] += aval * bt[j]
                j += 1
              end
            end
            t += 1
          end
          i += 1
        end
        out
      end

      def transpose(a)
        a[0].each_index.map { |j| a.map { |row| row[j] } }
      end

      # row-wise add of a bias vector
      def add_bias(a, bias)
        a.map { |row| row.each_index.map { |j| row[j] + bias[j] } }
      end

      def add(a, b)
        a.each_index.map { |i| a[i].each_index.map { |j| a[i][j] + b[i][j] } }
      end

      def scale(a, s)
        a.map { |row| row.map { |v| v * s } }
      end

      # numerically stable softmax over a row vector
      def softmax(vec)
        mx = vec.max
        exps = vec.map { |v| Math.exp(v - mx) }
        sum = exps.sum
        sum = 1e-12 if sum.zero?
        exps.map { |e| e / sum }
      end

      # layer normalization of a row vector
      def layernorm(vec, gamma: nil, beta: nil, eps: 1e-5)
        mean = vec.sum / vec.length
        var = vec.sum { |v| (v - mean)**2 } / vec.length
        inv = 1.0 / Math.sqrt(var + eps)
        vec.each_index.map do |i|
          y = (vec[i] - mean) * inv
          g = gamma ? gamma[i] : 1.0
          b = beta ? beta[i] : 0.0
          y * g + b
        end
      end

      def gelu(x)
        0.5 * x * (1.0 + Math.tanh(Math.sqrt(2.0 / Math::PI) * (x + 0.044715 * x**3)))
      end

      def gelu_mat(a)
        a.map { |row| row.map { |v| gelu(v) } }
      end

      def cosine(u, v)
        dot = 0.0
        nu = 0.0
        nv = 0.0
        u.each_index do |i|
          dot += u[i] * v[i]
          nu += u[i]**2
          nv += v[i]**2
        end
        d = Math.sqrt(nu) * Math.sqrt(nv)
        d.zero? ? 0.0 : dot / d
      end

      # A tiny, seedable xorshift64 PRNG (deterministic, no Kernel#rand global).
      class PRNG
        def initialize(seed)
          @s = (seed.to_i & 0xFFFFFFFFFFFFFFFF)
          @s = 0x9E3779B97F4A7C15 if @s.zero?
        end

        def next_u64
          x = @s
          x ^= (x << 13) & 0xFFFFFFFFFFFFFFFF
          x ^= (x >> 7)
          x ^= (x << 17) & 0xFFFFFFFFFFFFFFFF
          @s = x & 0xFFFFFFFFFFFFFFFF
          @s
        end

        # float in [0,1)
        def next_float
          (next_u64 >> 11) / (1 << 53).to_f
        end
      end
    end
  end
end

File: apps/core/src/main/java/bada/mind/Tensor.java
package bada.mind;

/**
 * Minimal pure-Java tensor helpers (matrices as double[][]) and a seeded
 * xorshift64 PRNG. Java port of Bada::Transformer::Tensor.
 */
public final class Tensor {

    public static double[][] matmul(double[][] a, double[][] b) {
        int m = a.length, k = b.length, n = b[0].length;
        double[][] out = new double[m][n];
        for (int i = 0; i < m; i++) {
            for (int t = 0; t < k; t++) {
                double av = a[i][t];
                if (av == 0.0) continue;
                double[] bt = b[t];
                double[] oi = out[i];
                for (int j = 0; j < n; j++) oi[j] += av * bt[j];
            }
        }
        return out;
    }

    public static double[][] transpose(double[][] a) {
        int r = a.length, c = a[0].length;
        double[][] t = new double[c][r];
        for (int i = 0; i < r; i++) for (int j = 0; j < c; j++) t[j][i] = a[i][j];
        return t;
    }

    public static double[][] add(double[][] a, double[][] b) {
        double[][] o = new double[a.length][a[0].length];
        for (int i = 0; i < a.length; i++)
            for (int j = 0; j < a[0].length; j++) o[i][j] = a[i][j] + b[i][j];
        return o;
    }

    public static double[][] addBias(double[][] a, double[] bias) {
        double[][] o = new double[a.length][a[0].length];
        for (int i = 0; i < a.length; i++)
            for (int j = 0; j < a[0].length; j++) o[i][j] = a[i][j] + bias[j];
        return o;
    }

    public static double[] softmax(double[] v) {
        double mx = Double.NEGATIVE_INFINITY;
        for (double x : v) mx = Math.max(mx, x);
        double sum = 0.0;
        double[] e = new double[v.length];
        for (int i = 0; i < v.length; i++) { e[i] = Math.exp(v[i] - mx); sum += e[i]; }
        if (sum == 0.0) sum = 1e-12;
        for (int i = 0; i < e.length; i++) e[i] /= sum;
        return e;
    }

    public static double[] layernorm(double[] v) {
        double mean = 0.0;
        for (double x : v) mean += x;
        mean /= v.length;
        double var = 0.0;
        for (double x : v) var += (x - mean) * (x - mean);
        var /= v.length;
        double inv = 1.0 / Math.sqrt(var + 1e-5);
        double[] o = new double[v.length];
        for (int i = 0; i < v.length; i++) o[i] = (v[i] - mean) * inv;
        return o;
    }

    public static double gelu(double x) {
        return 0.5 * x * (1.0 + Math.tanh(Math.sqrt(2.0 / Math.PI) * (x + 0.044715 * x * x * x)));
    }

    public static double[][] geluMat(double[][] a) {
        double[][] o = new double[a.length][a[0].length];
        for (int i = 0; i < a.length; i++)
            for (int j = 0; j < a[0].length; j++) o[i][j] = gelu(a[i][j]);
        return o;
    }

    public static double cosine(double[] u, double[] v) {
        double dot = 0, nu = 0, nv = 0;
        for (int i = 0; i < u.length; i++) { dot += u[i] * v[i]; nu += u[i] * u[i]; nv += v[i] * v[i]; }
        double d = Math.sqrt(nu) * Math.sqrt(nv);
        return d == 0 ? 0.0 : dot / d;
    }

    public static double[][] randmat(int rows, int cols, PRNG rng, double scale) {
        double[][] m = new double[rows][cols];
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++) m[i][j] = (rng.nextFloat() * 2.0 - 1.0) * scale;
        return m;
    }

    /** Deterministic xorshift64 PRNG. */
    public static final class PRNG {
        private long s;
        public PRNG(long seed) {
            s = seed;
            if (s == 0) s = 0x9E3779B97F4A7C15L;
        }
        public long nextU64() {
            long x = s;
            x ^= (x << 13);
            x ^= (x >>> 7);
            x ^= (x << 17);
            s = x;
            return x;
        }
        public double nextFloat() {
            return (nextU64() >>> 11) / (double) (1L << 53);
        }
    }

    private Tensor() { }
}
